### PR TITLE
Implemented TWAP oracle

### LINKED ISSUE #1163: TWAP based oracle in AMM
https://docs.uniswap.org/protocol/V2/concepts/core-concepts/oracles

https://github.com/Uniswap/v2-core/blob/master/contracts/UniswapV2Pair.sol#L72-L86

https://github.com/Uniswap/v2-periphery/blob/master/contracts/examples/ExampleOracleSimple.sol

https://github.com/Uniswap/v2-periphery/blob/master/contracts/libraries/UniswapV2OracleLibrary.sol#L16-L35

### PR DESCRIPTION
**Pull request by ashi20192020ayush**
resolves #1163

### CODE REVIEW

File: pallets/amm/src/lib.rs [hunk 1]
```diff
@@ -610,10 +613,46 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
...
+
+            let price0_fraction: BalanceOf<T, I> =
+                FixedU128::saturating_from_rational(pool.quote_amount, pool.base_amount)
+                    .into_inner();
+            let price1_fraction: BalanceOf<T, I> =
+                FixedU128::saturating_from_rational(pool.base_amount, pool.quote_amount)
+                    .into_inner();
+
+            pool.price_0_cumulative_last =
+                pool.price_0_cumulative_last.saturating_add(price0_fraction);
```
Comment (GopherJ) [inline pallets/amm/src/lib.rs:635]: not good to have so many saturating I think, can we find another way to make it not overflow?

Comment (ashi20192020ayush) [reply pallets/amm/src/lib.rs:635]: let me try @GopherJ

**Comment by drbh** [comment]
minor updates

1. added test
2.  moved `pool.block_timestamp_last = block_timestamp;` to bottom of the `update_oracle` function

**Review by drbh — APPROVED** [review]
LGTM for auditing purposes, we can optimize `saturating_mul`s later

**Pull request merged**